Fix large list truncation

hithomasmorelli opened the pull request:
Fixes #126 

- [x] Include tests for bug fix and new functionality
- [ ] Updated documentation for new feature (n/a)

hithomasmorelli commented:
@browniebroke I've fixed the bug and updated the tests (to check for item truncation where applicable), but I'm having trouble with the [cassettes](https://github.com/browniebroke/deezer-python/tree/master/tests/cassettes) - `tox -e lint,docs,py38` is failing on `pytest` because new requests are being made (to test for truncation) and I'm getting multiple `vcr.errors.CannotOverwriteExistingCassetteException`s.

How should I solve this? Apologies for my lack of knowledge - to say I'm rusty with `pyvcr` would be an understatement, I've never used it before!

hithomasmorelli commented:
OK, so I think I've fixed the issue of not being able to overwrite existing cassettes (by first running `pytest --vcr-record new_episodes`), but I've got a different sort of issue now:

I added an extra test to `TestPlaylist.test_get_fans`, in order to test that the list of fans isn't being truncated (apart from at 100, where the API `fans` endpoint has a firm limit), however that generated a cassette file with the IDs and usernames of 100 fans of a playlist I randomly chose.

I feel a bit conflicted regarding having this username data within this repository - the Deezer website doesn't provide functionality to view this data, it seems to only be available via the API. Additionally, while there might be a reasonable expectation that Deezer editors might have their track/album/artist/playlist details extracted from the API (given their roles), the 100 fans that the API spits out haven't had an opportunity to consent to their IDs & usernames being stored in a random repository.

Because of this, I propose that, before I commit & push the changes, I manually edit the `TestPlaylist.test_get_fans` cassette file to blank out the IDs & usernames of the new users being retrieved. This shouldn't affect the truncation test as it's just checking for the length of the list.
@browniebroke is this OK with you?

browniebroke commented:
Hey, thanks for taking some time on this. Yes, sounds reasonable to anonymise the data before pushing it. 👍

hithomasmorelli commented:
Brilliant, will push now

browniebroke commented on deezer/client.py:
```diff
@@ -185,6 +185,27 @@ def get_object(
+        # fix deezer's truncation of items (e.g. tracks on large playlists)
+        if (
+            (
+                relation == "tracks"
+                or relation == "fans"
+                or relation == "albums"
+                or relation == "artists"
+                or relation == "playlists"
```
Why add these restrictions on specific relations? Could we could do it whenever there is a `next` in the response, regardless of the relation being requested?

hithomasmorelli replied:
I remember testing it without that and something going horribly wrong, but that was before `and limit not in kwargs` got inserted. I'll do some messing around and see if I can repro the same issue again, otherwise I'll take them out

hithomasmorelli replied:
Still can't remember exactly why I left that in, but I think it had something to do with endpoints such as (and possibly only) `artist/<artist>/top`. I suppose I thought it was a different question whether artists' top songs should be listed beyond the API's default limit (without specifying a higher limit).

The benefits of following "next" URLs for requesting (e.g.) playlist tracks is fairly clear-cut in my opinion, as - if you ask for the tracks of a playlist - you expect all the tracks from that playlist to be returned. It's a bit more unclear as to whether all of an artist's top 100 songs (the API maximum) should be returned by default when making that call.

browniebroke commented on deezer/client.py:
```diff
@@ -185,6 +185,27 @@ def get_object(
+            and "limit" not in kwargs
+        ):
+            new_json = json
+            while "next" in new_json:
+                response = self.session.get(new_json["next"])
+                new_json = response.json()
+                json["data"].extend(new_json["data"])
```
This is bypassing the error handling that we have above. What happens if we get an error in the middle? Could we get throttled by Deezer in this section?

hithomasmorelli replied:
I'll fix this. What do you suggest happens if we get throttled? My justification for fixing this in the first place was that "if you ask for `<playlist>`'s tracks, you want *all* the tracks", but what should happen if we get cut off halfway through (or whenever)?

browniebroke commented on .gitignore:
```diff
@@ -18,3 +18,6 @@ ENV*
 yarn.lock
+
+# Visual Studio Code
+.vscode
```
I don't mind keeping this, but FYI editor specific files should usually be added in your own global `.gitignore`:

https://gist.github.com/subfuzion/db7f57fff2fb6998a16c

hithomasmorelli replied:
Thanks - wasn't aware

hithomasmorelli commented:
> Looks good overall, I'm just wondering if this feature could trigger a large amount of hidden API calls. Did you hit any issues with that?

I did think about that, but I'm not sure if there is a better (feasible) solution. I ran into this issue because I was requesting tracks from a large playlist and they weren't all appearing, which seemed contrary to the design of the library. The Deezer API's preferred way of doing things is seemingly to follow `next` URLs (I couldn't find any docs on these, it's just [what the API does](https://stackoverflow.com/a/22654992/1831866)), so it seems to me like the current best solution is to do this. Yes, it will lead to more hidden API calls, but the alternative (not providing all items in response to a request for them) doesn't seem logical imo.

Just to say, when making changes to the tests I did originally have tests where 1000-1500 songs were requested and I don't remember hitting any rate limits (the API does have some but I can't find them formally documented, just this [uncited answer](https://stackoverflow.com/a/28278457/18318660)) - I changed those tests to reduce the size of the cassettes. Arguably, if the API throws a rate limit for following its own `next` URLs in an attempt to get a complete list, that's a fault of the API itself 🙃

hithomasmorelli commented:
I've removed `.vscode` from .gitignore, and added error handling to the `next` requests.

To avoid duplicate code, I merged the `next` requests functionality with the original request functionality. Only consequence of this is that the code is a pain to read unaided - hence the barrage of comments.

If you know a better way to code this, feel free to [suggest an] edit. I am definitely not claiming this is the best way of doing things - it's just the only one that came to mind!

Not quite sure why the bandit lint check failed, running `bandit -r deezer` on my machine results in 0 issues.

hithomasmorelli commented:
Just discovered while working on #134 that `Resource.iter_relation` does this in a more managed way (so as not to use loads of memory at once). Entirely my fault that I hadn't seen this sooner - I was (wrongly) focusing on the `get_*`, and not the `iter_*`.

https://github.com/browniebroke/deezer-python/blob/99f01b79d91811d69e3b95a89ea8c38c84f2b404/deezer/resources.py#L65-L72

I'm still of the opinion that if you ask for the tracks in a playlist (in list form), you expect *all* the tracks from that playlist. But now the code changes basically render any memory usage reduction in `Resource.iter_relation` useless, as `Client.get_object` will return everything at once anyway.

Hmm. I'm a bit stumped now as to what's the best way to implement this. Any help from anyone would be greatly appreciated!

browniebroke commented:
I haven't got around to review this again properly yet, but yes, the `iter_...` methods are basically doing a similar thing in a different way.

Do you think we could stick to the `iter_...` constructs for now? That would probably make the change fit better in the current design.

We might want to switch the `get_...` to have a better automatic handling of pagination, but that seems like a bigger, potentially breaking change. One option would be to switch them to do what `iter_...` are doing right now, and essentially merge the 2 together. 

I recently used [PyGithub](https://github.com/PyGithub/PyGithub) and I liked how they [handle pagination](https://pygithub.readthedocs.io/en/latest/utilities.html#pagination), it works out of the box and transparently goes through the pages while also letting you get a specific page or the total count of elements in the result set.

hithomasmorelli commented:
Yes, that’s probably a good idea. I’ll close this PR for now (should I leave #126 open?)?

 

 

From: Bruno Alla <notifications@github.com>
Reply to: browniebroke/deezer-python <reply@reply.github.com>
Date: Wednesday, 19 August 2020 at 08:16
To: browniebroke/deezer-python <deezer-python@noreply.github.com>
Cc: Thomas Morelli <me@thomasmorelli.me>, Author <author@noreply.github.com>
Subject: Re: [browniebroke/deezer-python] Fix large list truncation (#127)

 

I haven't got around to review this again properly yet, but yes, the iter_... methods are basically doing a similar thing in a different way.

Do you think we could stick to the iter_... constructs for now? That would probably make the change fit better in the current design.

We might want to switch the get_... to have a better automatic handling of pagination, but that seems like a bigger, potentially breaking change. One option would be to switch them to do what iter_... are doing right now, and essentially merge the 2 together.

I recently used PyGithub and I liked how they handle pagination, it works out of the box and transparently goes through the pages while also letting you get a specific page or the total count of elements in the result set.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

browniebroke commented:
I think it could be closed yes. I'll write an issue to rework the pagination, the fact that you missed the `iter_...` APIs is a sign that it could be improved.

browniebroke commented:
Opened #137
